### DISCUSSION TITLE
Using custom Bootstrap JS modules

### DISCUSSION THREAD

**Discussion by axlright** [holger1411/understrap#1287] [General]
Hello friends-

Are any of you compiling custom / selected Bootstrap JS modules instead of the full `bootstrap.js`? If so, mind sharing your setup? I have some clients who need super compiled / reduced everything for performance reasons.

Looking at `node_modules/bootstrap/js` there's individual modules in `dist` and in `src` directories but I'm not sure the difference between them.

I'd assume a workflow would be to modify `gulpfile.js` to pull all those `node_modules/bootstrap/js` and then modify the `scripts` function to compile certain components in place of `paths.dev + '/js/bootstrap4/bootstrap.js',`

Any input is appreciated!

-Alex

**Comment by 0dp** [comment]
hey Buddy, hope you are doing well!

Imma have a look and get right back to you

**Comment by 0dp** [comment]
There are two options right of the bat.

## Quick and dirty
Just like you write, you can import each file from `node_modules/bootstrap/js/dist` as you need them...You always need `util.js`

How you chose to do that depends on how complicated you want it, the out come will be the same.

Start by changing the copy-assets task to copy `node_modules/bootstrap/js/dist` files instead of just the bootstrap bundle and copy all the files into js folder.
```
	var stream = gulp
		.src(`${paths.node}bootstrap/js/dist/**/*.js`)
		.pipe(gulp.dest(`./js/bootstrap4`));
```

create a bootstrap.js file in ./src called boostrap for instance.
```
import './bootstrap4/util.js'
import './bootstrap4/alert.js'
import './bootstrap4/button.js';
...
```

Then modify the scripts task

```
`${paths.dev}/js/bootstrap.js`,
```

Now when you run scripts task these files will be imported, unminified, as is, no thrills.
Pretty much the same as importing them directly from the boostrap npm package. Janky af 👎 

To get a better performance and minification. You could include every single module into the scripts task instead of importing them into one file. 

## The right way
The right way to do this would be to setup a propper build system for this, that would take care of these thing automatically. Which you know is something I've been pushing for, for years now.

Basically create a build system that pulls in all the dependencies from our modules in compile time instead of the copy-assets task or scripts task.

This is what you see most modern apps do these days. I was about to go on a rant about this, but I'll leave it here. IMHO to do this properly it would get political real quick, since it would introduce breaking changes on so many levels.

**Comment by Thomas-A-Reinert** [comment]
I am totally with you. It loads (possible) bloat.
MY personally preferred way would be an installer script that asks within initial "`npm install`" which js-modules you want. According to that it would write out some config-file, letting the gulp-task know which modules to load/compile.
Same goes for bootstrap.css by the way. I almost never use tables and such, so it would be really great to have some installer script/dialog about which css-modules you want to integrate and get rid of the rest.

**Comment by axlright** [comment]
Thanks @0dp and @Thomas-A-Reinert !!

As always, y'all are more talented than I am. I'm gonna craft up and test a more quick and dirty approach and I'll post the workflow here and let y'all know the results.

**Comment by 0dp** [comment]
It should work just fine buddy 💪

**Comment by axlright** [comment]
I tested with this in `gulpfile.js` as an edit to the `gulp scripts` action:

```
gulp.task( 'scripts', function() {
	var scripts = [

		// Start Bootstrap
		// Optionally use all Bootstrap JS components...
		// `${paths.dev}/js/bootstrap4/bootstrap.bundle.js`,

		//... or comment out above and select only components that the project uses while always including util.js

		`${paths.dev}/js/bootstrap4/util.js`, // always include
		`${paths.dev}/js/bootstrap4/alert.js`, //NAME
		`${paths.dev}/js/bootstrap4/button.js`, //NAME$1
		`${paths.dev}/js/bootstrap4/carousel.js`, //NAME$2
		`${paths.dev}/js/bootstrap4/collapse.js`, //NAME$3
		`${paths.dev}/js/bootstrap4/dropdown.js`, //NAME$4
		`${paths.dev}/js/bootstrap4/modal.js`, // NAME$5
		`${paths.dev}/js/bootstrap4/tooltip.js`, //NAME$6
		`${paths.dev}/js/bootstrap4/popover.js`, //NAME$7
		`${paths.dev}/js/bootstrap4/scrollspy.js`, //NAME$8
		`${paths.dev}/js/bootstrap4/tab.js`, //NAME$9
		`${paths.dev}/js/bootstrap4/toast.js`, // NAME$a
		`${paths.dev}/js/bootstrap4/index.js`,
		// End Bootstrap

		`${paths.dev}/js/skip-link-focus-fix.js`,

		// START CUSTOMIZED ADDITIONS - PRESERVE THROUGH UPDATES
		// Owl Carousel 2 scripts
		// paths.dev + '/js/owl.carousel.js',

		// Custom scripts
		paths.custom_js + '*.js'

		// END CUSTOMIZED ADDITIONS - PRESERVE THROUGH UPDATES
	];
  gulp.src( scripts, { allowEmpty: true } )
	.pipe( concat( 'app.min.js' ) )
	.pipe( uglify() )
	.pipe( gulp.dest( paths.js ) );

  return gulp.src( scripts, { allowEmpty: true } )
	.pipe( concat( 'app.js' ) )
	.pipe( gulp.dest( paths.js ) );
});
```

The js file it creates is smaller than that of the one using `bootstrap.bundle.js` and it throws a console error for two instances of "Uncaught TypeError: Bootstrap's JavaScript requires jQuery. jQuery must be included before Bootstrap's JavaScript." which come from this in `index.js` though everything (in my tests so far) work including collapse, dropdown and a few others.

```
if (typeof $ === 'undefined') {
    throw new TypeError('Bootstrap\'s JavaScript requires jQuery. jQuery must be included before Bootstrap\'s JavaScript.');
  }
```

Commenting out `${paths.dev}/js/bootstrap4/index.js,` makes other things unhappy.

There's not a ton of information on forums / elsewhere online about compiling a custom set of Bootstrap js components. Any thoughts? I'm gonna keep experimenting.

**Comment by 0dp** [comment]
Well the most important thing is you include the utils.js
Whatever you include after that should work. That's what I did.

I agree there is not a ton of info about this.
But try with one module at the time.
Like modal for instance. Then build on that.

You got this

**Comment by axlright** [comment]
Thanks @0dp here's what worked for me - removing the `index.js` after got rid of all console errors...

```
// Start Bootstrap
        // Optionally use all Bootstrap JS components...
        // `${paths.dev}/js/bootstrap4/bootstrap.bundle.js`,

        //... or comment out above and select only components that the project uses while always including util.js

        `${paths.dev}/js/bootstrap4/util.js`, // always include
        `${paths.dev}/js/bootstrap4/alert.js`, //NAME
        `${paths.dev}/js/bootstrap4/button.js`, //NAME$1
        `${paths.dev}/js/bootstrap4/carousel.js`, //NAME$2
        `${paths.dev}/js/bootstrap4/collapse.js`, //NAME$3
        `${paths.dev}/js/bootstrap4/dropdown.js`, //NAME$4
        `${paths.dev}/js/bootstrap4/modal.js`, // NAME$5
        `${paths.dev}/js/bootstrap4/tooltip.js`, //NAME$6
        `${paths.dev}/js/bootstrap4/popover.js`, //NAME$7
        `${paths.dev}/js/bootstrap4/scrollspy.js`, //NAME$8
        `${paths.dev}/js/bootstrap4/tab.js`, //NAME$9
        `${paths.dev}/js/bootstrap4/toast.js`, // NAME$a
        // `${paths.dev}/js/bootstrap4/index.js`,
        // End Bootstrap
```
I'll keep working project by project and maybe snoop around the Bootstrap Github to see if there's more explanation of how to compile individual modules.

**Comment by Thomas-A-Reinert** [comment]
> Thanks @0dp and @Thomas-A-Reinert !!
> 
> As always, y'all are more talented than I am.

Sorry to say, but that´s total BS ;)
I usually have a problem, no clue about how to solve it but dig into it like a mole and don´t let go like a stafford terrier. YOU solved this problem within hours, didn´t you? Don´t hide your light under a bushel, dude..

**Comment by mitchelldmiller** [comment]
Hello Nice People:
I learned programming on punch cards, never learned new NPM stuff. But this is how I got example by @axlright to work on my system. Copied Babel() from original file.
Best wishes,
Mitchell

       gulp.task( 'scripts', function() {
    	 var scripts = [
    		paths.node + '/bootstrap/js/src/util.js',
     		paths.node + '/bootstrap/js/src/alert.js',
	     	paths.node + '/bootstrap/js/src/button.js',
     		paths.node + '/bootstrap/js/src/carousel.js',
    		paths.node + '/bootstrap/js/src/collapse.js',
    		paths.node + '/bootstrap/js/src/dropdown.js',
   		paths.node + '/bootstrap/js/src/modal.js',
   		paths.node + '/bootstrap/js/src/tooltip.js',
   	   	paths.src + '/js/skip-link-focus-fix.js',
         	];
   
     gulp.src( scripts, { allowEmpty: true } )
    	.pipe( babel( { presets: ['@babel/preset-env'] } ) )
       	.pipe( concat( 'app.min.js' ) )
     	.pipe( uglify() )
      	.pipe( gulp.dest( paths.js ) );
     
       return gulp.src( scripts, { allowEmpty: true } )
     	.pipe( babel() )
     	.pipe( concat( 'app.js' ) )
     	.pipe( gulp.dest( paths.js ) );
    });

**Reply by 0dp** [reply]
@danemorgan and crew.

What would you prefer:

To have a gulp scripts modified so you can uncomment BS js files you don't need?

Or include webpack-stream into the scripts task, then have a js file in src where you can uncomment BS js files you don't need?

Same same but different.

I like webpack-stream personally because it will allow for all the webpack goodness like lazy loading, splitting etc in the future.

**Reply by mitchelldmiller** [reply]
Gulp scripts, because I understand them.

**Reply by danemorgan** [reply]
Good point @mitchelldmiller.

I think more people who use this theme are likely familiar with Gulp than with Webpack?

But would adding something like this from Webpack allow us to ease into a long-term plan to migrate to a more modern build system, in a manageable way @0dp? Or would it be better to make a clean break when the time comes?

**Reply by 0dp** [reply]
Yes it would.

To easy your mind @mitchelldmiller the actual webpack bundling in our case would be extremely shallow. You'd barely notice it in the gulp scripts task.
`.pipe(webpack({}));`

That's it!!

Configuring bootstrap JS files to include is merely a matter of commenting them in a js file in src. Conveniently named bootstrap.js
Just like we have with scss.

So in that way it also aligns with an existing 'pattern'

**Comment by bacoords** [comment]
Interesting..

So the main question is- where should we be responsible for loading the bootstrap js modules, in case someone wants to turn some off. When I did this recently for a child theme, I did it in the gulpfile by separating them out, which would be my default answer.

Assuming I understand correctly, if we go the webpack method, that's more like using 'require' at the top of our source JS file, meaning we'd have a child-theme.js file where you could comment out modules and add your custom JS below, similar to a typical SCSS set up. Do I have that right?

My questions are:
1. Will it cause any bloat for the end user - or will the final JS filesize be the same?
2. Are there any other dependencies needed for developers or any weirdness that makes this any 'harder' for a new dev to be able to spin up?

**Reply by 0dp** [reply]
1. Bloat, no not really. Ultimately it's the same as using bootstrap.bundle.js bit you can choose to make it less bloaty.
2. No not at all. Either method, it will all be embedded in the scripts task so nothing breaking there. In terms of dependencies it will be two extra npm modules to install. Webpack-stream and popper.js

**Comment by danemorgan** [comment]
Hi @0dp 

I'm going to be transparent as I can with you.

All of this kind of tooling is new to me. I started using Gulp about 2 years ago, but I did that using a package from Ahmad Awais that I just add to my theme folder. It worked and did what was on the tin and I never looked deeper. I've only used Bootstrap lightly on HDC projects and they've always been already set up and I just had to work with the SCSS.

I've never even looked at Webpack. Everything you're doing is beyond me with this.

One question I have is doesn't Webpack to some extent replace or duplicate what Gulp does? If we went to Webpack for part of the build process, would we benefit from going fully to Webpak, or do Webpack and Gulp each have their own domain?

**Reply by 0dp** [reply]
Yes and no.
Gulp is a task runner. It runs tasks.
Webpack is a bundler. It bundles stuff, but can also run tasks. So on a very simple level webpack is superior.

But it's not forbidden to run webpack in gulp in order to bundle stuff.

Though in the future, I'd like to see gulp replaced by a more modern build system.

**Comment by 0dp** [comment]
Aside, i think what ever solution we pick. It's a valuable performance improvement 👍

**Reply by bacoords** [reply]
I was looking at what they're doing at Bootstrap, which I think should be our default path to follow if we were to change anything. 

I'm not fully sure but it looks like they're using a bunch of custom npm scripts, mixed with postcss for scripts and babel for JS. 

I'd be interested in learning more about their method, though I personally see no major reason to move away from the currenty gulp/babel workflow which is working fine currently.

**Reply by 0dp** [reply]
They use rollup for JavaScript on 4.6.

But changing that will be a breaking change and very different from what we are doing now.

**Comment by Thomas-A-Reinert** [comment]
I´ll just repeat myself. And it ain´t rocket science to achieve that.

> I am totally with you. It loads (possible) bloat.
> MY personally preferred way would be an installer script that asks within initial "npm install" which js-modules you want. According to that it would write out some config-file, letting the gulp-task know which modules to load/compile.
> Same goes for bootstrap.css by the way. I almost never use tables and such, so it would be really great to have some installer script/dialog about which css-modules you want to integrate and get rid of the rest.

For me that looks like the most idiot-proof solution. Remember - there are DEVS using understrap and well.. "others" who don´t know what they´re doing anyway, who just install that theme because it looks nice and appealing to them somehow. That would get the latter on board too.  At least maybe. >

I wonder, how you do not know the differences between webpack, gulp, grunt, rollup, vite etc. Thought you were some up-to-date web-dev-agency. Gulp seems like some legacy thing to me nowadays.  And NO: I´m totally against jumping pon the latest bandwagon and have any working tools adjusted towards that latest trend. But Webpack ain´t "new" at all anymore. Gulp was already outdated like two years ago compared to other solutions.

Webpack can do whatever gulp does, plus a lot more and that a LOT faster and better. Gulp became a no go for me within the past two years at least. Yes it works. But it takes time. A LOT compared to Webpack depending on which version you want to use.

However - the question is if you want to IMPROVE things for the better on a somehow "outdated" version of bootstrap/understrap and if you wanna stick to the flow it had since it´s beginning. The former workflow should be outdated by adapting BS5 anyway for the better.

On the pro side:
- Perfomance improvement galore
- Still working with IE11 for ppl who need to support it ( Well, the integration of Webpack alone will still support IE11. BS5 won´t )

On the con side:
- Some ppl will need to dig into new territory
- Not that this cannot be documented - but they´re old well-known command will prolly not work anymore. I could live with that. RTFM for a new version, I´d say. If you miss those, well that´s your personal problem. But I know there will be a whole lot of questions showing up on the issue tracker. Which could be answered with a simple text-module.
